core/src/main/java/org/infinispan/transaction/xa/recovery/RecoveryManagerImpl.java

```diff
@@ -128,38 +130,43 @@ public RecoveryIterator getPreparedTransactionsFromCluster() {
    }
 
    @Override
-   public void removeRecoveryInformation(Collection<Address> lockOwners, Xid xid, boolean sync, GlobalTransaction gtx,
+   public CompletionStage<Void> removeRecoveryInformation(Collection<Address> lockOwners, Xid xid, GlobalTransaction gtx,
                                          boolean fromCluster) {
       log.tracef("Forgetting tx information for %s", gtx);
       //todo make sure this gets broad casted or at least flushed
       if (rpcManager != null && !fromCluster) {
          TxCompletionNotificationCommand ftc = commandFactory.buildTxCompletionNotificationCommand(xid, gtx);
-         sendTxCompletionNotification(lockOwners, ftc, sync);
+         return sendTxCompletionNotification(lockOwners, ftc)
+               .thenRun(() -> removeRecoveryInformation(xid));
```

Comment:
I think it's safe to remove the local recovery info in the caller thread and then return the stage.

Slightly off-topic, I've been looking at `RecoveryAwareTransactionTable.removeRemoteTransaction` and it seems horribly inefficient. It should only iterate over the transactions when there the `GlobalTransaction` is not available.

Reply:
I can try that.